### PR TITLE
Added sarif template

### PR DESCRIPTION
**Pull request by rahul2393**
[Issue](https://github.com/aquasecurity/trivy/issues/547)

### CODE REVIEW

**Comment by rahul2393** [comment]
@knqyf263 kindle review

**Comment by knqyf263** [comment]
@simar7 Could you review this PR? We need to upload a SARIF file to GitHub actually and check how it looks on the UI.

File: contrib/sarif.tpl [hunk 1]
```diff
@@ -0,0 +1,80 @@
+{
+  "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
+  "version": "2.1.0",
+  "runs": [
+    {
+      "tool": {
+        "driver": {
+          "name": "Trivy: Vulnerability Scanner for Containers",
+          "semanticVersion": "0.9.1",
```
Comment (simar7) [inline contrib/sarif.tpl:9]: We should make this dynamic rather than hard coding it to a particular version.

**Comment by simar7** [comment]
Is this valid SARIF? For me it doesn't pass the validator checks. You can try this here https://sarifweb.azurewebsites.net/Validation

image

I think you could also use a simple JSON validator as it should pass that too.

**Comment by simar7** [comment]
Furthermore, could you also add a test that would exercise this template? You can see an example here https://github.com/aquasecurity/trivy/blob/aa20adb22be4839c3dca76c842470b288e434220/integration/client_server_test.go#L78-L87

**Comment by rahul2393** [comment]
@simar7 fixed all comments, also validated the final output in Sarif validator
Screenshot 2020-07-16 at 9 17 57 AM

File: contrib/sarif.tpl [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+              },
+              "fullDescription": {
+                "text": {{ .Description | printf "%q" }}
+              },
+              "defaultConfiguration": null
+              "properties": {
+                "tags": [
+                  "{{ .PkgName }}"
+                ],
+                "precision": "very-high"
```
[the remark this answers is not in the document]
Comment (rahul2393) [reply contrib/sarif.tpl:32]: This means how accurate is result we need to set it "very-high"

File: integration/client_server_test.go [hunk 3]
```diff
@@ -464,7 +476,7 @@ func setupClient(t *testing.T, c args, addr string, cacheDir string, golden stri
 
 	// Setup the output file
 	var outputFile string
-	if *update {
+	if *update || createNewFile {
```
Comment (simar7) [inline integration/client_server_test.go:479]: Why do we need `createNewFile`? The existing test already creates a temp file for the output that we can use to compare.

File: pkg/report/writer.go [hunk 4]
```diff
@@ -52,6 +52,12 @@ func WriteResults(format string, output io.Writer, results Results, outputTempla
 				}
 				return escaped.String()
 			},
+			"endWithPeriod": func(input string) string {
+				if !strings.HasSuffix(input, ".") {
+					input += "."
+				}
+				return input
+			},
```
Comment (simar7) [inline pkg/report/writer.go:60]: This seems untested. You can add a unit test here https://github.com/aquasecurity/trivy/blob/e7e3bb45ce59b736e29a88cbe68a6970db241bdf/pkg/report/writer_test.go#L199-L294

Comment (rahul2393) [reply pkg/report/writer.go:60]: @simar7 Done

**Pull request merged**